compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeCompatibilityChecker.kt

```diff
@@ -0,0 +1,489 @@
+                if ((ctx.prohibitComparisonOfIncompatibleEnums && typeParameterOwner.symbol.classId == StandardClassIds.Enum) ||
+                    (ctx.prohibitComparisonOfIncompatibleClasses && typeParameterOwner.symbol.classId == StandardClassIds.KClass)
+                ) {
+                    compatibilityUpperBound
+                } else {
+                    Compatibility.SOFT_INCOMPATIBLE
+                }
+            BoundTypeArguments(mutableSetOf(), mutableSetOf(), compatibilityUpperBoundForTypeArg)
+        }.let {
+            val type = boundTypeArgument.type
+            if (boundTypeArgument.variance.allowsInPosition) {
+                it.lower += type.collectLowerBounds()
+            }
+            if (boundTypeArgument.variance.allowsOutPosition) {
+                it.upper += type.collectUpperBounds()
+            }
+        }
+    }
+
+    private fun FirClassLikeDeclaration<*>.getSuperTypes(): List<ConeClassLikeType> {
+        return when (this) {
+            is FirTypeAlias -> listOfNotNull(expandedTypeRef.coneTypeSafe())
+            is FirClass<*> -> superTypeRefs.mapNotNull { it.coneTypeSafe() }
+            else -> emptyList()
+        }
+    }
+
+    private fun ConeClassLikeType.getClassLikeElement(ctx: ConeInferenceContext): FirClassLikeDeclaration<*>? =
+        ctx.symbolProvider.getSymbolByLookupTag(lookupTag)?.fir
+
+    private fun FirClassLikeDeclaration<*>.getTypeParameter(index: Int): FirTypeParameterRef? {
+        return when (this) {
+            is FirTypeAlias -> typeParameters[index]
+            is FirClass<*> -> typeParameters[index]
+            else -> return null
+        }
+    }
+
+    /** A class declaration and the arguments bound to the declared type parameters. */
+    private data class TypeArgumentMapping(
+        val typeParameterOwner: FirClassLikeDeclaration<*>,
+        val mapping: Map<FirTypeParameterRef, BoundTypeArgument>
+    )
+
+    /** A single bound type argument to a type parameter declared in a class. */
+    private data class BoundTypeArgument(val type: ConeKotlinType, val variance: Variance)
+
+    /** Accumulated type arguments bound to a type parameter declared in a class. */
+    private data class BoundTypeArguments(
+        val upper: MutableSet<ConeClassLikeType>,
+        val lower: MutableSet<ConeClassLikeType>,
+        val compatibilityUpperBound: Compatibility
+    )
+
+    private fun ConeClassLikeType.toConcreteClass(ctx: ConeInferenceContext): ConcreteClass? {
+        return lookupTag.toConcreteClass(ctx)
+    }
+
+    private fun ConeClassLikeLookupTag.toConcreteClass(
+        ctx: ConeInferenceContext
+    ): ConcreteClass? = when (val klass = ctx.symbolProvider.getSymbolByLookupTag(this)?.fir) {
+        is FirTypeAlias -> klass.getAliasedClass(ctx)?.let { ConcreteClass.TypeAlias(klass, it, ctx) }
+        is FirClass<*> -> ConcreteClass.Class(klass, ctx)
+        else -> null
+    }
+
+    private sealed class ConcreteClass {
```

Comment:
I'm not sure this class is needed. You can use `fullyExpandedType()` to unfold all alias-based types.

Reply:
Thanks! Yep I don't need the sealed class. But I still need a wrapper so that locaitng super classes can be lazy. Let me make this a simple wrapper instead.